### PR TITLE
[GET] Implement retrieve one entry endpoint + service + repo + unit tests

### PR DESCRIPTION
**Pull request by hrafov**

### CODE REVIEW

File: META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
+Main-Class: com.example.statussvc.Application
```
Comment (yurii-polishchuk) [inline META-INF/MANIFEST.MF]: Please remove this file

File: src/main/java/com/example/statussvc/controller/HostsController.java [hunk 2]
```diff
@@ -16,7 +19,6 @@
 @RequiredArgsConstructor
 @RequestMapping(API_V1)
 public class HostsController {
-
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/controller/HostsController.java:32]: please leave this new line as was before.

File: src/main/java/com/example/statussvc/controller/HostsController.java [hunk 3]
```diff
@@ -31,8 +33,9 @@ public Object modify() {
         return null;
     }
 
-    public Object retrieve() {
-        return null;
+    @GetMapping("/host/{id}")
+    public Host retrieve(@PathVariable Long id) {
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/controller/HostsController.java]: Please refactor.

1. use custom RetrieveHostResponse and create it in a wire package.
2. use wrapper ResponseEntity

`public ResponseEntity<RetrieveHostResponse> retrieve(@PathVariable Long id) { ... }`

File: src/main/java/com/example/statussvc/controller/HostsController.java [hunk 4]
```diff
@@ -31,8 +33,9 @@ public Object modify() {
...
+        return HostsRepository.getHost(id);
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/controller/HostsController.java]: do not call Repository in Controller. Please read Spring MVC + Patterns:

- https://www.baeldung.com/spring-mvc
- https://tom-collings.medium.com/controller-service-repository-16e29a4684e5

File: src/main/java/com/example/statussvc/domain/Host.java [hunk 5]
```diff
@@ -23,5 +23,13 @@ public class Host implements Domain {
     @Id
     @GeneratedValue
     private Long id;
+    private String title;
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/domain/Host.java]: please remove. should be created by @AntonTsa

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.web.servlet.MockMvc;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+@ExtendWith(SpringExtension.class)
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: new line after import

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+@WebMvcTest(controllers = RootController.class)
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: why are you testing Root Controller?

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+@AutoConfigureMockMvc
+@ActiveProfiles("test")
+@MockBean(Tracer.class)
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class HostControllerTest {
+    private final MockMvc mockMvc;
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java:51]: new line after {

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    private final Tracer tracer;
+    @BeforeEach
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: new line after tracer;

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        given(span.context())
+                .willReturn(TraceContext.newBuilder()
+                        .traceId(1)
+                        .spanId(1)
+                        .build()
+                );
+    }
+    @Test
+    @DisplayName("""
+            GIVEN valid root endpoint
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please revisit Display Name and read about it. https://mkyong.com/junit5/junit-5-display-names/

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                        .content(StringUtils.EMPTY))
+
+                // THEN
+                .andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        // AND THEN
+        assertThat(!actualResponse.getContentAsString().isEmpty());
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: do not use ! in asserts. Please read the documentation and check how to assert.

plus, why are you testing `/hosts` ? your endpoint is `/hosts/{id}`

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    @DisplayName("""
+            GIVEN valid root endpoint
+            WHEN performing valid GET request
+            THEN return empty success response
+            """)
+    void HostEndpointNotValid() throws Exception {
+        // GIVEN
+
+        // WHEN
+        MockHttpServletResponse actualResponse = mockMvc.perform(get("/abc")
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: what exactly are you testing here? we don't have `/abc` endpoint

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                        .content(StringUtils.EMPTY))
+
+                // THEN
+                .andExpect(status().isNotFound())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        // AND THEN
+        assertThat(actualResponse.getContentAsString().isEmpty());
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: what your test is testing?

File: src/main/java/com/example/statussvc/controller/HostsController.java [hunk 14]
```diff
@@ -31,8 +38,10 @@ public Object modify() {
         return null;
     }
 
-    public Object retrieve() {
-        return null;
+    @GetMapping("/hosts/{id}")
+    public ResponseEntity<URL> retrieve(@PathVariable("id") Long id) throws URISyntaxException {
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/controller/HostsController.java]: Please read: https://www.restapitutorial.com/lessons/httpmethods.html and update properly Controller based on REST Specification.

File: src/main/java/com/example/statussvc/domain/Host.java [hunk 15]
```diff
@@ -1,27 +1,39 @@
...
 
 /**
  * Main Host domain biz object
  */
 @Getter
 @Setter
 @ToString
-@Table
+@Data
+@Table(name = "URLS")
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/domain/Host.java]: why URLS? table should be named hosts as per mapping

File: src/main/java/com/example/statussvc/service/HostsService.java [hunk 16]
```diff
@@ -24,8 +28,11 @@ public Object modify() {
         return null;
     }
 
-    public Object retrieve() {
-        return null;
+    public Host retrieve(Long id) {
+
+        HostsRepository repository;
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/service/HostsService.java]: use @Autowired to inject repository

File: src/main/resources/db.migration.mysql/V1__init.sql [hunk 17]
```diff
@@ -0,0 +1,11 @@
+Create table urls (
```
Comment (yurii-polishchuk) [inline src/main/resources/db.migration.mysql/V1__init.sql]: please remove - this is not a part of your PR

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+                        .build()
+                );
+    }
+    @Test
+    @DisplayName("""
+            GIVEN valid endpoint
+            WHEN performing valid GET request
+            THEN return success response as list of all hosts
+            """)
+    void ValidHostsEndpoint() throws Exception {
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please remove this test, and read carefully your task.

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+            GIVEN valid endpoint
+            WHEN performing valid GET request
+            THEN return success response
+            """)
+    void HostEndpointWithValidId() throws Exception {
+        this.mockMvc
+                .perform(get("/hosts/1", 1))
+                .andDo(print()).andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect((ResultMatcher) jsonPath("$.id").value("1"));
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please use ObjectMapper to get data from the response and compare via assert with your prepared expected results

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    @DisplayName("""
+            GIVEN invalid endpoint
+            WHEN performing valid GET request
+            THEN return Not Found response
+            """)
+    void HostEndpointNotValid() throws Exception {
+        // GIVEN
+
+        // WHEN
+        MockHttpServletResponse actualResponse = mockMvc.perform(get("/bad_hosts")
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: what exactly are you testing here? that there is no endpoint mapped? this test is not helping to test your /hosts/{id} endpoint at all

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+                .content(StringUtils.EMPTY))
+
+                // THEN
+                .andExpect(status().isNotFound())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+                .andReturn()
+                .getResponse();
+
+        // AND THEN
+        assertThat(actualResponse.getStatus() == 404);
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please read how to properly use asserts in ANY kind of testing.

- https://www.baeldung.com/introduction-to-assertj
- https://www.tutorialspoint.com/junit/junit_using_assertion.htm
- https://www.guru99.com/junit-assert.html
- https://www.vogella.com/tutorials/AssertJ/article.html

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+    @DisplayName("""
+            GIVEN valid endpoint
+            WHEN performing not valid GET request
+            THEN return Bad Request response
+            """)
+    void HRequestIsNotValid() throws Exception {
+        // GIVEN
+
+        // WHEN
+        MockHttpServletResponse actualResponse = mockMvc.perform(get("/hosts?")
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: again, your not testing your functionality, just some random endpoints - that are not bringing any value by a test

**Comment by AntonTsa** [comment]
1) Names of columns in Entity Host don't match ones in sql script
2) Repository is not initialized in Service
3) Incorrect Controller's logic. Implemented POST method instead of GET

**Review by yurii-polishchuk — CHANGES_REQUESTED** [review]

File: META-INF/MANIFEST.MF (hunk 1 above)
Comment (yurii-polishchuk) [reply META-INF/MANIFEST.MF]: still a valid comment, please remove.

File: src/main/java/com/example/statussvc/controller/HostsController.java [hunk 23]
```diff
@@ -31,8 +36,9 @@ public Object modify() {
         return null;
     }
 
-    public Object retrieve() {
-        return null;
+    @GetMapping("/host/{id}")
+    public ResponseEntity<RetrieveHostResponse> retrieve(@PathVariable Long id) {
+        return RetrieveHostResponse.findById(id);
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/controller/HostsController.java]: you need to call Service layer to e.g. `hostService.retrieve(id)`

File: src/main/java/com/example/statussvc/domain/type/Host.java [hunk 24]
```diff
@@ -0,0 +1,4 @@
+package com.example.statussvc.domain.type;
+
+public class Host {
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/domain/type/Host.java]: please remove this class, we already have Host.java object

File: src/main/java/com/example/statussvc/service/HostsService.java [hunk 25]
```diff
@@ -12,6 +15,9 @@
 @RequiredArgsConstructor
 public class HostsService {
 
+    @Autowired
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/service/HostsService.java]: annotation is not needed here. just:
```
private final HostsRepository hostsRepository;
```

File: src/main/java/com/example/statussvc/service/HostsService.java [hunk 26]
```diff
@@ -24,8 +30,8 @@ public Object modify() {
         return null;
     }
 
-    public Object retrieve() {
-        return null;
+    public Host retrieve(Long id) {
+        return hostsRepository.findById(id).orElseThrow();
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/service/HostsService.java]: orElseThrow() what? I'd recommend throwing 404 e.g. ```HttpClientErrorException.NotFound```

File: src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java [hunk 27]
```diff
@@ -0,0 +1,6 @@
+package com.example.statussvc.wire.response;
+
+import com.example.statussvc.wire.Response;
+
+public class RetrieveHostResponse implements Response {
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java:8]: please add fields to response + I would suggest to use `record` from Java16+

File: src/main/resources/db.migration.mysql/V1__init.sql (hunk 17 above)
Comment (yurii-polishchuk) [reply src/main/resources/db.migration.mysql/V1__init.sql]: still valid comment, please remove

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+        given(tracer.currentSpan()).willReturn(span);
+        given(span.context())
+                .willReturn(TraceContext.newBuilder()
+                        .traceId(1)
+                        .spanId(1)
+                        .build()
+                );
+    }
+    @Test
+    @DisplayName("get endpoint with existing id")
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please use GIVEN / WHEN / THEN

**Review by yurii-polishchuk — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/statussvc/service/HostsService.java [hunk 29]
```diff
@@ -31,7 +29,12 @@ public Object modify() {
     }
 
     public Host retrieve(Long id) {
-        return hostsRepository.findById(id).orElseThrow();
+        try {
+            return hostsRepository.findById(id)
+                    .orElseThrow(() -> new HttpClientErrorNotFoundException("Host not found"));
+        } catch (HttpClientErrorNotFoundException e) {
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/service/HostsService.java]: why are you throwing and then immediately catching an exception? o_O

File: src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java [hunk 30]
```diff
@@ -1,6 +1,13 @@
 package com.example.statussvc.wire.response;
 
 import com.example.statussvc.wire.Response;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Date;
 
-public class RetrieveHostResponse implements Response {
+public record RetrieveHostResponse(@JsonProperty String title,
```
Comment (yurii-polishchuk) [inline src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java]: can we avoid @JsonProperty ?

File: src/test/java/com/example/statussvc/controller/HostControllerTest.java [hunk 31]
```diff
@@ -51,7 +52,7 @@ public void mockTracer() {
                 );
     }
     @Test
-    @DisplayName("get endpoint with existing id")
+    @DisplayName("get endpoint with existing id - expected 200 OK")
```
Comment (yurii-polishchuk) [inline src/test/java/com/example/statussvc/controller/HostControllerTest.java]: please check format of the test and display name naming convention

```
"""
GIVEN ...
WHEN ...
THEN ...
"""
```

**Review by yurii-polishchuk — CHANGES_REQUESTED** [review]

File: META-INF/MANIFEST.MF (hunk 1 above)
Comment (yurii-polishchuk) [reply META-INF/MANIFEST.MF]: still valid, please remove

File: gitignore [hunk 32]
```diff
@@ -0,0 +1 @@
+MANIFEST.MF
```
Comment (yurii-polishchuk) [inline gitignore]: we already have gitignore file, please reuse

File: src/main/java/com/example/statussvc/domain/type/Host.java (hunk 24 above)
Comment (yurii-polishchuk) [reply src/main/java/com/example/statussvc/domain/type/Host.java]: please pull latest changes

File: src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java (hunk 27 above)
Comment (yurii-polishchuk) [reply src/main/java/com/example/statussvc/wire/response/RetrieveHostResponse.java:8]: any updates here?

File: src/main/resources/application.yml [hunk 33]
```diff
@@ -8,9 +8,10 @@ spring:
     bannerMode: off
   # Data source
   dataSource:
-    url: test
-    username: test
-    password: test
+    driverClassName: com.mysql.jdbc.Driver
```
Comment (yurii-polishchuk) [inline src/main/resources/application.yml]: pool latest changes from main

File: src/test/resources/application-test.yml [hunk 34]
```diff
@@ -4,9 +4,10 @@ spring:
     bannerMode: console
   # Data source
   dataSource:
-    url: test
-    username: test
-    password: test
+    driverClassName: com.mysql.jdbc.Driver
```
Comment (yurii-polishchuk) [inline src/test/resources/application-test.yml]: pull latest changes, this was fixed.

**Review by yurii-polishchuk — CHANGES_REQUESTED** [review]

**Review by yurii-polishchuk — APPROVED** [review]

**Review by yurii-polishchuk — DISMISSED** [review]

**Pull request closed without merging**